README.md
# Semantic Table

## NEW! React 18 Support

DataTable made with styled components from [React Semantic UI](https://react.semantic-ui.com).

Installation:

```
npm i semantic-table
```

Dependencies:
This component depends on the following packages

- [React Semantic UI](https://react.semantic-ui.com).
- [Cleave JS](https://nosir.github.io/cleave.js/).
- [React CSV](https://github.com/react-csv/react-csv).
- [Moment JS](https://momentjs.com)

## Basic Example

---

```
const data = [
    {
        id: 9888,
        name: "Hartmann, Runolfsson and Stroman Inc",
        email: "Moshe67@gmail.com"
    },
    {
        id: 28451,
        name: "Jacobson Inc Inc",
        email: "Green19@hotmail.com"
    },
    {
        id: 72265,
        name: "Bahringer, Padberg and Upton Inc",
        email: "Casandra_Heaney97@gmail.com"
    },
]

<PaginatedTable
    title="Example"
    onCancel={true}
    rows={data}
    rowLimit={15}
    actionsActive={false}
    columns={[
        {
            width: 250,
            Header: 'Code',
            accessor: 'id',
        },
        {
            width: 225,
            Header: 'Name',
            accessor: 'name',
        },
        {
            width: 225,
            Header: 'Email',
            accessor: 'email',
        },
    ]}
/>
```

## Props

---

| Prop                       | Type               | Description                                                                                                                                                                                                                                                                                                                             |
| -------------------------- | ------------------ | --------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| title                      | string             | The title used for exporting table data to CSV                                                                                                                                                                                                                                                                                          |
| rows                       | array              | Array of objects to be displayed in the table                                                                                                                                                                                                                                                                                           |
| columns                    | array              | Array of colums to be displayed, each column will access to the corresponding attribute from each row data (see structure [here](#columns)) it can control whether the column can be filtered, ordered and edited                                                                                                                       |
| rowLimit                   | numeric            | The number of rows to be displayed per page                                                                                                                                                                                                                                                                                             |
| loading                    | boolean            | Controls whether the table will show a "loading" message to indicate that data is being loaded into the table                                                                                                                                                                                                                           |
| actionsActive              | boolean            | Whether each row will have a set of actions (see structure [here](#row-actions))                                                                                                                                                                                                                                                        |
| actionsWidth               | numeric            | Width in pixels of the first column that will display the actions buttons                                                                                                                                                                                                                                                               |
| onSave                     | function           | Function to execute to save changes on rows                                                                                                                                                                                                                                                                                             |
| enableExternalSave         | boolean            | Controls whether table should save via external manipulation                                                                                                                                                                                                                                                                            |
| onAdd                      | function           | Function to execute if the user wants to create a new item with an external component (form or modal form)                                                                                                                                                                                                                              |
| onAddRow                   | boolean / function | If set to true, it will create a new row on click, if a function is given, it will expect an object with custom data to match the given rows structure                                                                                                                                                                                  |
| onSelect                   | function           | Function to execute if the user clicks on the "pencil" icon button on a specific row it as access to the selected row                                                                                                                                                                                                                   |
| onDelete                   | boolean / function | If set to true, it will display a button to remove the selected row ("trash" icon button), if a function is given, it will allow a custom function to be executed after the deletion has been made, the function has access to the remaining rows, if the function returns a false boolean, then it will not trigger the inner deletion |
| onCancel                   | boolean / function | If set to true, it will cancel all changes made on the table, if a function is given, it will allow a custom function to be executed after the cancelation has been made                                                                                                                                                                |
| onEditCell                 | function           | Function to execute after a cell has been edited, it will execute on blur, if the function returns a false boolean, then it will not trigger the inner edition                                                                                                                                                                          |
| onInnerUpdate              | function           | Function to execute after any table rows change                                                                                                                                                                                                                                                                                         |
| enableExportToCSV          | boolean            | If set to true, it will display a button that will export all displayed data in the table (ignoring pagination, it will only export data that hasn't been filtered) in a Comma Separated Values (CSV) format, it uses the "title" prop and a timestamp to name the exported file                                                        |
| height                     | numeric            | The table height, max varies depending on screen size                                                                                                                                                                                                                                                                                   |
| additionalActionButtons    | array              | Additional buttons to be displayed on each row, each button will have a custom action and icon to give it the desired look and functionality (see structure [here](#columns))                                                                                                                                                           |
| additionalHeaderButtons    | array              | Additional buttons to be displayed on the header section, each button will have a custom action and icon to give it the desired look and functionality (see structure [here](#additional-header-actions))                                                                                                                               |
| saveButtonText             | string             | Custom text for save button                                                                                                                                                                                                                                                                                                             |
| cancelButtonText           | string             | Custom text for cancel button                                                                                                                                                                                                                                                                                                           |
| newRowButtonText           | string             | Custom text for newRow button                                                                                                                                                                                                                                                                                                           |
| addButtonText              | string             | Custom text for add button                                                                                                                                                                                                                                                                                                              |
| removeFiltersButtonText    | string             | Custom text for removeFilters button                                                                                                                                                                                                                                                                                                    |
| hideRemoveFiltersButton    | boolean            | Control whether the removeFiltersButton will appear or not                                                                                                                                                                                                                                                                              |
| editRowButtonText          | string             | Custom text for editRow ave button                                                                                                                                                                                                                                                                                                      |
| deleteRowButtonText        | string             | Custom text for deleteRow button                                                                                                                                                                                                                                                                                                        |
| exportToCSVButtonText      | string             | Custom text for exportToCSV button                                                                                                                                                                                                                                                                                                      |
| actionsHeaderText          | string             | Custom text for actions header                                                                                                                                                                                                                                                                                                          |
| numberOfRecordsText        | string             | Custom text for total records label                                                                                                                                                                                                                                                                                                     |
| showRecords                | boolean            | Control whether the table will show total of records                                                                                                                                                                                                                                                                                    |
| numberOfRecordsPerPageText | string             | Custom text for total records per page label                                                                                                                                                                                                                                                                                            |
| showRecordsPerPage         | boolean            | Controls whether the table will show total of records per page                                                                                                                                                                                                                                                                          |
| paginated                  | boolean            | Controls whether the table is paginated or not, defaults to true                                                                                                                                                                                                                                                                        |

## Loading

---

```
const data = [
    {
        id: 9888,
        name: "Hartmann, Runolfsson and Stroman Inc",
        email: "Moshe67@gmail.com"
    },
    {
        id: 28451,
        name: "Jacobson Inc Inc",
        email: "Green19@hotmail.com"
    },
    {
        id: 72265,
        name: "Bahringer, Padberg and Upton Inc",
        email: "Casandra_Heaney97@gmail.com"
    },
]

const [isLoading, setIsLoading] = useState(false)

useEffect(() => {
    setIsLoading(true)
    setTimeout(() => {
        setIsLoading(false)
    }, 1500)
}, [])


<PaginatedTable
    loading={isLoading}
    title="Example"
    onCancel={true}
    rows={data}
    rowLimit={15}
    actionsActive={false}
    columns={[
        {
            width: 250,
            Header: 'Code',
            accessor: 'id',
        },
        {
            width: 225,
            Header: 'Name',
            accessor: 'name',
        },
        {
            width: 225,
            Header: 'Email',
            accessor: 'email',
        },
    ]}
/>
```

## Header Actions

---

| Prop              | Type               | Description                                                                                                                                                                                                                                                                      |
| ----------------- | ------------------ | -------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| onSave            | function           | Function to execute to save changes on rows                                                                                                                                                                                                                                      |
| onAdd             | function           | Function to execute if the user wants to create a new item with an external component (form or modal form)                                                                                                                                                                       |
| onAddRow          | boolean / function | If set to true, it will create a new row on click, if a function is given, it will expect an object with custom data to match the given rows structure                                                                                                                           |
| onCancel          | boolean / function | If set to true, it will cancel all changes made on the table, if a function is given, it will allow a custom function to be executed after the cancelation has been made                                                                                                         |
| onEditCell        | function           | Function to execute after a cell has been edited, it will execute on blur                                                                                                                                                                                                        |
| enableExportToCSV | boolean            | If set to true, it will display a button that will export all displayed data in the table (ignoring pagination, it will only export data that hasn't been filtered) in a Comma Separated Values (CSV) format, it uses the "title" prop and a timestamp to name the exported file |

### Example

```
const data = [
    {
        id: 9888,
        name: "Hartmann, Runolfsson and Stroman Inc",
        email: "Moshe67@gmail.com"
    },
    {
        id: 28451,
        name: "Jacobson Inc Inc",
        email: "Green19@hotmail.com"
    },
    {
        id: 72265,
        name: "Bahringer, Padberg and Upton Inc",
        email: "Casandra_Heaney97@gmail.com"
    },
]


<PaginatedTable
    title="Example"
    onCancel={true}
    rows={data}
    rowLimit={15}
    actionsActive={false}
    onSave={({ deleteItems, updateItems, createItems, rows }) => {
        console.log(deleteItems)
        console.log(updateItems)
        console.log(createItems)
        console.log(rows)
    }}
    onAdd={() => {
        console.log('This can be used to display an external form or modal form! Control the rendered rows externally')
    }}
    onAddRow={(id) => {
        return {
            id: id,
            name: "Daugherty Inc LLC"
            email: "Connor.Bergstrom56@yahoo.com"
        }
    }}
    onCancel={true}
    onEditCell={(editedRow) => {
        console.log('This is the new data: ')
        console.log(editedRow)
    }}
    enableExportToCSV={true}
    columns={[
        {
            width: 250,
            Header: 'Code',
            accessor: 'id',
        },
        {
            width: 225,
            Header: 'Name',
            accessor: 'name',
        },
        {
            width: 225,
            Header: 'Email',
            accessor: 'email',
        },
    ]}
/>
```

## Row Actions

---

| Prop                    | Type               | Description                                                                                                                                                                                                                                       |
| ----------------------- | ------------------ | ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| actionsActive           | boolean            | Whether each row will have a set of actions (see structure [here]())                                                                                                                                                                              |
| actionsWidth            | numeric            | Width in pixels of the first column that will display the actions buttons                                                                                                                                                                         |
| onSelect                | function           | Function to execute if the user clicks on the "pencil" icon button on a specific row it as access to the selected row                                                                                                                             |
| onDelete                | boolean / function | If set to true, it will display a button to remove the selected row ("trash" icon button), if a function is given, it will allow a custom function to be executed after the deletion has been made, the function has access to the remaining rows |
| additionalActionButtons | array              | Additional buttons to be displayed on each row, each button will have a custom action and icon to give it the desired look and functionality, icons [here](https://react.semantic-ui.com/elements/icon/)                                          |

### Example

```
const data = [
    {
        id: 9888,
        name: "Hartmann, Runolfsson and Stroman Inc",
        email: "Moshe67@gmail.com"
    },
    {
        id: 28451,
        name: "Jacobson Inc Inc",
        email: "Green19@hotmail.com"
    },
    {
        id: 72265,
        name: "Bahringer, Padberg and Upton Inc",
        email: "Casandra_Heaney97@gmail.com"
    },
]

const additionalActionButtons = [
    {
        name: 'View',
        icon: 'eye',
        action: (row) => {
            console.log(row)
        }
    },
     {
        name: 'Email',
        icon: 'mail',
        action: (row) => {
            console.log('Mail sent to ' + row.email)
        }
    }
]

<PaginatedTable
    title="Example"
    onCancel={true}
    rows={data}
    rowLimit={15}
    actionsActive={true}
    actionsWidth={125}
    onSelect={(selectedRow) => {
        console.log(selectedRow)
    }}
    onDelete={(remainingRows) => {
        console.log(remainingRows)
    }}
    additionalActionButtons={additionalActionButtons}
    columns={[
        {
            width: 250,
            Header: 'Code',
            accessor: 'id',
        },
        {
            width: 225,
            Header: 'Name',
            accessor: 'name',
        },
        {
            width: 225,
            Header: 'Email',
            accessor: 'email',
        },
    ]}
/>
```

## Columns

Each column will have the following structure:

| Prop            | Type    | Description                                                                                                                        |
| --------------- | ------- | ---------------------------------------------------------------------------------------------------------------------------------- |
| width           | numeric | The column's width, data exceeding this width will trigger an horizontal scrollbar                                                 |
| Header          | string  | The column's title to be displayed on the table header                                                                             |
| accessor        | string  | The name of the row attribute that holds the value to be displayed                                                                 |
| type            | string  | Format and input type that the column will have, can be one of: **text, currency, boolean, select, textarea, number, color, date** |
| options         | array   | If the type is **select**, then an options array has to be provided (**see the structure below**)                                  |
| editable        | boolean | Whether the column can be edited or not                                                                                            |
| filterable      | boolean | Whether the column can be filtered or not                                                                                          |
| permanentRender | boolean | Whether the column cell input will always be visible and not render the default label                                              |

### Column Options

| Prop  | Type              | Description                                                                                                                                                                                                   |
| ----- | ----------------- | ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |

CHANGELOG.md
## Major Version v1.0.1 Latest (2022-07-17)

### Features
- Added support for React 18

### Dependencies
- React 18

### Bugs

## v0.3.8 (2022-06-2)
---

### Features

### Dependencies

### Bugs
- Fixed ColumnCard filtering for values not on first page
- Fixed true boolean values not being parsed correctly

## v0.3.7 (2022-06-2)
---
**[a7dd716](https://github.com/MarilesBeard5/semantic-table/commit/a7dd716aa8dc46de6001c63e1ac2e32a2408ff0b)**

### Features

### Dependencies

### Bugs
- Fixed ColumnCard filtering for nested attributes of select type

## v0.3.6 (2022-06-1)
---
**[a7dd716](https://github.com/MarilesBeard5/semantic-table/commit/a7dd716aa8dc46de6001c63e1ac2e32a2408ff0b)**

### Features
- Added **permanentRender** prop to allow perpetual rendering of input on cells, usage should be use with caution as could lead to
  performance issues

### Dependencies

### Bugs

## v0.3.5 (2022-05-31)
---
**[9337d8f](https://github.com/MarilesBeard5/semantic-table/commit/9337d8f7d12a16f4d072707828d12a9b56df41ed)**
### Features
- Added row data to additionalActionButtons to conditionally render as a function, if a functions is given, it will expect the callback to return the
  actual array of options

### Dependencies

### Bugs

## v0.3.4 (2022-05-26)
---
**[a10815f](https://github.com/MarilesBeard5/semantic-table/commit/a10815fae875af2ea59f0ecef3a35165784bd52b)**

**[d821cbb](https://github.com/MarilesBeard5/semantic-table/commit/d821cbb2d379e0c686a5f598ef3a4f005def2a6b)**

**[ca58c99](https://github.com/MarilesBeard5/semantic-table/commit/ca58c99454ccae4c2bca572e851dac9dd7802c43)**

### Features
- Added Pagination Component to fix previous pagination bugs and improve stability
- Added external control functions to avoid inner re-rendering and state update, this is to allow controlled updating of table rows

### Dependencies
- REMOVED [Decimal JS](https://mikemcl.github.io/decimal.js/)

### Bugs
* Fixed card pointing direction
* Changed container width calculation to automatic

## v0.3.2 (2022-05-24)
---
**[0bd8be9](https://github.com/MarilesBeard5/semantic-table/commit/0bd8be94da4545d1200e899780159ef33f49bde7)**

### Features
* Added decimal operations to allow better table width handling

### Dependencies
- [Decimal JS](https://mikemcl.github.io/decimal.js/)
### Bugs
* Fixed boolean cell rendering on edition

## v0.2.8 (2022-05-16)
---
**[a1396e8](https://github.com/MarilesBeard5/semantic-table/commit/a1396e8fc26af599ce86efcfb273305028f28e0b)**

**[153a720](https://github.com/MarilesBeard5/semantic-table/commit/153a720a68e461cc65e15b51824a064425d99c2c)**

**[0185567](https://github.com/MarilesBeard5/semantic-table/commit/01855676e234500906de33fea8996c7cf4c10362)**
### Features
* Added additional header action buttons
* Added records per page label
* Added prop to allow saving on external manipulation

### Dependencies

### Bugs
* Fixed rows to identify edition on additional action save
* Fixed logical rendering of filter card options through checked property to allow external controlling

## v0.2.6 (2022-04-28)
---
**[b194e31](https://github.com/MarilesBeard5/semantic-table/commit/b194e310b8c8e5c9a3c7c1fa35bfe979dafee8cc)**

### Features
* Added better support for date filtering
* Added hideRemoveFilterButton prop and showTotalRecords prop
* Added better uuid identifier to avoid collisions

### Dependencies

### Bugs
* Fixed onBlur bug where rows where being replaced
* Fixed unpaginated table filtering issue
* Fixed boolean edition
* Fixed filtering card positioning

## v0.2.4 (2022-04-20)
---
**[288c287](https://github.com/MarilesBeard5/semantic-table/commit/288c287e711553cb9984cfe499fc64e69ac51a92)**

**[20dad95](https://github.com/MarilesBeard5/semantic-table/commit/20dad95531fd60072bb231c00ce5b84cb4964caf)**

**[f60f042](https://github.com/MarilesBeard5/semantic-table/commit/f60f04220516a733cce76012317661439546161e)**

### Features
* Added faker JS to be available for cloned repos
* Removed automatic sorting, now sort will be applied on pressing the "Aplicar" button
* Added custom props for spanish text labels
* Added more data to onEditCell and onDelete functions
* Added support for static date formatting to allow better filtering
* Added total rows label

### Dependencies

### Bugs

## v0.2.1 (2022-04-19)
---
**[7daf5aa](https://github.com/MarilesBeard5/semantic-table/commit/7daf5aaa5f1e6040005f3795627b59475e1b03c9)**

**[2b87a5d](https://github.com/MarilesBeard5/semantic-table/commit/2b87a5d166bc0409df842c29c1766ec1c0a1b7b7)**

### Features
* Added custom height prop for table
* Added tab indexing for each editable column
* Added better processing functions on utils file
* Added faker for basic local tests
* Increased max height of table
* Added unshift functionality for new rows instead of pushing to the end
* Added different cursor to assert where a user can edit
* Added better documentation
* Added color change on focus a row and pagination items

### Dependencies
* Faker JS

### Bugs
* Fixed _select_ type column filtering
* Fixed onCancel function bug
* Fixed key generation for additional columns
* Fixed repeated re-rendering for checking if table can be saved or not
* Fixed specific column focusing to change overflowing on specific cells and not all same-type cells.

## v0.1.8 (2022-04-12)
---
**[9609911](https://github.com/MarilesBeard5/semantic-table/commit/960991176c5f3c18d047d0eb1b6448fb5fdce706)**

**[96e4f1f](https://github.com/MarilesBeard5/semantic-table/commit/96e4f1f486953b51489b871ee2bc2660fa42d49b)**

**[14e2645](https://github.com/MarilesBeard5/semantic-table/commit/14e264569a1cc7ba2076e3a5c7a1721759816b64)**

**[04336c8](https://github.com/MarilesBeard5/semantic-table/commit/04336c8a82f73c4fc8852c47d8a0ee561f7e07c7)**

**[cda125a](https://github.com/MarilesBeard5/semantic-table/commit/cda125a21a326ab3f1ff9ed3a73b6e1b06b0e0b6)**

### Features
* Added _add new row_ functionality
* Added paginated boolean option to hide pagination item

### Dependencies

### Bugs

src/App.js
import { faker } from '@faker-js/faker'
import { useEffect, useState, useMemo } from 'react'
import { Button, Segment, Tab } from 'semantic-ui-react'
import './index.css'
import PaginatedTable from './lib/components/PaginatedTable'
import { generateRows } from './lib/utils/fakes'

const clientsSchema = {
	id: '{{datatype.number}}',
	name: '{{company.companyName}} {{company.companySuffix}}',
	email: '{{internet.email}}',
	date: '{{date.between}}',
	processed: '{{datatype.boolean}}',
}

const options = [
	{
		text: 'Foo',
		value: '01',
	},
	{
		text: 'Bar',
		value: '02',
	},
]

function App() {
	const [isLoading, setIsLoading] = useState(false)
	const [data, setData] = useState(generateRows(350, 560, clientsSchema))
	const [canSave, setCanSave] = useState(false)

	useEffect(() => {
		setIsLoading(true)
		setTimeout(() => {
			setIsLoading(false)
		}, 1500)
	}, [data])

	const memoizedData = useMemo(() => {
		return (data ?? []).map(d => {
			return {
				...d,
				type: '01'
			}
		})
	}, [data])

	const panes = [
		{
			menuItem: 'Tab 1',
			render: () => (
				<div className='PageContainer'>
					<Button
						content='Recargar Datos'
						type='button'
						onClick={() => {
							setData(generateRows(350, 560, clientsSchema))
							setCanSave(true)
						}}
					/>
					<PaginatedTable
						loading={isLoading}
						title='Example'
						hideRemoveFiltersButton={true}
						onCancel={() => {
							setCanSave(false)
						}}
						enableExternalSave={canSave}
						onSave={true}
						onAddRow={(id) => ({
							id: id,
							name: faker.fake(
								'{{company.companyName}} {{company.companySuffix}}'
							),
							email: faker.fake('{{internet.email}}'),
							date: faker.fake('{{date.between}}'),
						})}
						rows={memoizedData}
						onInnerUpdate={(rows) => {
							console.log(rows)
						}}
						rowLimit={25}
						actionsActive={true}
						actionsWidth={150}
						onSelect={(row) => {
							console.log(row)
						}}
						onDelete={(deletedRow) => {
							console.log(deletedRow)
						}}
						onEditCell={({ row, column, value }) => {
							console.log(row)
							console.log(column)
							console.log(value)
						}}
						additionalActionButtons={(row) => {
							if (row.processed === 'true') {
								return [
									{
										name: 'Prueba',
										icon: 'search',
										action: (row) => {
											console.log(row)
										},
									},
								]
							} else {
								return [
									
								]
							}
						}}
						columns={[
							{
								width: 250,
								Header: 'Code',
								accessor: 'id',
								filterable: true,
								editable: false,
							},
							{
								width: 225,
								Header: 'Name',
								accessor: 'name',
								filterable: true,
								editable: true,
							},
							{
								width: 225,
								Header: 'Email',
								accessor: 'email',
								filterable: true,
								editable: true,
							},
							{
								width: 225,
								Header: 'Type',
								accessor: 'type',
								filterable: true,
								editable: true,
								type: 'select',
								options: options,
							},
							{
								width: 225,
								Header: 'Date',
								accessor: 'date',
								type: 'date',
								filterable: true,
								editable: true,
							},
							{
								width: 225,
								Header: 'Checked',
								accessor: 'processed',
								type: 'boolean',
								filterable: true,
								editable: true,
								permanentRender: true,
							},
						]}
					/>
				</div>
			),
		},
		{
			menuItem: 'Tab 2',
			render: () => <></>,
		},
	]

	return (
		<div className='PageContainer'>
			<Segment secondary>
				<Tab panes={panes} menu={{ pointing: true, secondary: true }} />
			</Segment>
		</div>
	)
}

export default App
